python/cudf/cudf/core/dataframe.py

```diff
@@ -1856,36 +1872,53 @@ def rename(self, mapper=None, columns=None, copy=True, inplace=False):
         Notes
         -----
         Difference from pandas:
-          * Support axis='columns' only.
-          * Not supporting: index, level
+            * Not supporting: level
 
         Rename will not overwite column names. If a list with duplicates is
         passed, column names will be postfixed with a number.
         """
-        # Pandas defaults to using columns over mapper
-        if columns:
-            mapper = columns
+        if mapper:
+            if axis == 0:
+                index = mapper
+            elif axis == 1:
+                columns = mapper
 
         out = DataFrame(index=self.index)
-        if isinstance(mapper, Mapping):
+
+        if columns:
             postfix = 1
-            # It is possible for DataFrames with a MultiIndex columns object
-            # to have columns with the same name. The followig use of
-            # _cols.items and ("_1", "_2"... allows the use of
-            # rename in this case
-            for key, col in self._data.items():
-                if key in mapper:
-                    if mapper[key] in out.columns:
-                        out_column = mapper[key] + "_" + str(postfix)
-                        postfix += 1
+            if isinstance(columns, Mapping):
+                # It is possible for DataFrames with a MultiIndex columns
+                # object to have columns with the same name. The followig
+                # use of _cols.items and ("_1", "_2"... allows the use of
+                # rename in this case
+                for key, col in self._data.items():
+                    if key in columns:
+                        if columns[key] in out.columns:
+                            out_column = columns[key] + "_" + str(postfix)
+                            postfix += 1
+                        else:
+                            out_column = columns[key]
+                        out[out_column] = col
                     else:
-                        out_column = mapper[key]
-                    out[out_column] = col
-                else:
-                    out[key] = col
-        elif callable(mapper):
-            for col in self._data.names:
-                out[mapper(col)] = self[col]
+                        out[key] = col
+            elif callable(columns):
+                for col in self._data.names:
+                    out[columns(col)] = self[col]
+        else:
+            out._data = self._data
+
+        if index:
+            out_idx = list(self.index)
+            if isinstance(index, Mapping):
+                for key, value in enumerate(out_idx):
+                    if value in index.keys():
+                        out_idx[key] = index[value]
+            elif callable(index):
+                for key, value in enumerate(out_idx):
+                    if value in index.keys():
+                        out_idx[key] = index(index[value])
```

Comment:
Lets not worry about supporting callables yet for index renames. We can't run element by element in Python land here, which means we'd need to pass the callable through Numba to generate a UDF which is super non-trivial and error prone.